mgnify_pipelines_toolkit/analysis/amplicon/amplicon_utils.py

```diff
@@ -0,0 +1,183 @@
+
+    return read_count
+
+def build_cons_seq(cons_list, read_count, cons_threshold=0.80, do_not_include=[], counter=1):
+
+    cons_seq = ''
+    cons_confs = []
+
+    for count_dict in cons_list:
+        max_count = 0
+        cons_dict = defaultdict(float)
+
+        if counter in do_not_include:
+            counter += 1
+            cons_seq += 'N'
+            continue 
+        
+        for base, count in count_dict.items():
+            if base not in ('A', 'T', 'C', 'G'):
+                continue
+
+            cons_dict[base] = count/read_count
+            
+            if count > max_count:
+                max_count = count
+
+        counter += 1
+        
+        try:
+            max_prop = max_count/read_count
+
+            cons_bases = []
+            curr_prop = 0.0
+            sorted_cons_dict = dict(sorted(cons_dict.items(), key=lambda x:x[1], reverse=True))
+
+            for base, prop in sorted_cons_dict.items():
+                cons_bases.append(base)
+                curr_prop += prop
+                if curr_prop >= cons_threshold:
+                    break
+
+            cons_bases = sorted(cons_bases)
+
+            if len(cons_bases) == 1:
+                cons_seq += cons_bases[0]
+            else:
+                amb_string = ','.join(cons_bases)
+                amb_base = _AMBIGUOUS_BASES_DICT_REV[amb_string]
+                cons_seq += amb_base
+                
+        except ZeroDivisionError:
+            max_prop = 0.0
+
+        cons_confs.append(max_prop)
+
+
+    return cons_seq, cons_confs
+
+def primer_regex_query_builder(primer):
+
+    query = ''
+
+    for char in primer:
+        if char in ('A', 'C', 'T', 'G'):
+            query += char
+        else:
+            query += str(_AMBIGUOUS_BASES_DICT[char])
+
+    query = f"(.*{query}){{e<=1}}"
+
+    return query
+
+def build_mcp_cons_dict_list(mcp_count_dict, mcp_len):
+    """
+    Generate list of dictionaries of base conservation for mcp output (mcp_cons_list)
+    e.g. [{'A':0.9, 'C':0.1}, {'T':1.0}, ....] for every base position
+    """
+
+    mcp_cons_list = []
+
+    for i in range(mcp_len):
+        index_base_dict = defaultdict(int)
+        for mcp in mcp_count_dict.keys():
+            if len(mcp) < mcp_len:
+                continue
+            base = mcp[i]
+            index_base_dict[base] += mcp_count_dict[mcp]
+        mcp_cons_list.append(index_base_dict)
+    
+    return mcp_cons_list
+
+def fetch_mcp(fastq, prefix_len, start=1, rev=False):
+    """
+    Runs a the "find_most_common_prefixes.sh" script to generate the mcps from a fastq file.
+
+    Outputs dictionary containing counts for each generated MCP in the fastq.
+
+    """
+
+    start = str(start)
+    prefix_len = str(prefix_len)
+
+    # Check strand requested
+    if not rev:
+        rev = '0'
+    else:
+        rev = '1'
+
+    cmd = [
+        'bash',
+        './mgnify_pipelines_toolkit/analysis/amplicon/find_most_common_prefixes.sh',
```

Comment:
linking shell scripts tends to create code that is tricky to follow, do you think we can convert that shell script to python code?